fix(ci): flaky tests

vados-cosmonic opened the pull request:
## Feature or Problem
<!---
Briefly describe the reason for this pull request: the feature being added or problem being solved.
--->

This test fixes some tests that were flaky in CI and removes waits.

Tests have only become *slightly* faster (on the order of seconds).

Some additional fixes were made to the nix tests, but it's not clear why they're failing (now that all deps are present, it runs for 25 minutes...)

## Related Issues
<!--- 
Link to any issues or correlated pull requests that are related to this PR. For example, if this PR fixes an issue, link to that issue here.
--->

## Release Information
<!---
Clearly state the target release for this code. If there isn't a specific target version, you can state the `next` release, etc. 
--->

`next`

## Consumer Impact
<!---
Indicate the impact, if any, this change will have on other consumers, dependencies, or dependents. In other words, the "blast radius" of the impact of this change and what steps related projects may need to take in response to this.
--->

Developers can more readily trust failing test runs to mean actual regressions.

## Testing
<!---
Declare the testing information for this pull request
--->

<!---
Identify the platforms on which this code was built (include both OS and CPU architecture)
--->
Built on platform(s)
- [ ] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

<!---
Identify the platforms on which this code was tested (include both OS and CPU architecture)
--->
Tested on platform(s)
- [ ] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

### Unit Test(s)
<!---
Indicate if unit tests were added or modified, and if so, which ones 
--->

### Acceptance or Integration
<!---
Indicate any changes or additions to the acceptance or integration test suite 
--->

### Manual Verification
<!---
Mandatory. Indicate the steps that you took to verify that this pull request works 
--->

rvolosatovs commented on Cargo.toml:
```diff
@@ -148,3 +148,6 @@ weld-codegen = "0.7.0"
 clap_complete = "4.3.0"
+
+[profile.dev.build-override]
+opt-level = 3
```
Why is this necessary? Considering that this option can slow down builds quite a bit and may not necessarily provide any performance benefits, this definitely requires a comment.
Also, why is this only done in `dev`? Why would we care about performance in `dev` profile?

For reference https://doc.rust-lang.org/cargo/reference/profiles.html#opt-level

vados-cosmonic replied:
Ah this was a retest of this option from early on, it was one of the first things I tried locally -- will remove this!

- [ ] remove dev build override

connorsmith256 reviewed (dismissed):
LGTM! One item for consideration: it looks like we mostly call `wait_until_process_has_count` on `beam.smp` with a predicate where we check for a count of 0. In one case we check for `nats-server` with a count of 1. If these are the only reasonable use cases, I'd consider wrapping these blocks in functions like `wait_for_host_to_stop` and `wait_for_nats_to_start`. Then you could remove the comments on these blocks, since the function call is self-describing :)

vados-cosmonic commented:
Hey @connorsmith256 yeah that's reasonable, will add a helper function that wraps that re-used functionality at a higher level. 

One thing I don't want to do is to make it seem like we're actually waiting for the host in an intelligent way though -- might make people think we're *not* just grepping for processes (essentially), since we don't know the PIDs to wait on specifically necessarily.

While I'm in here I'm going to mark the macOS nix build as optional since sometimes it OOMs (and there's not much we can do about that right now)

vados-cosmonic commented:
One more rebase (I refactored and made the nix MacOS builds optional for now) -- @connorsmith256 if you wouldn't mind taking another look

rvolosatovs commented on .github/workflows/nix.yml:
```diff
@@ -67,10 +67,12 @@ jobs:
     name: wash-${{ matrix.config.target }}
     runs-on: macos-12
+    # NOTE(vados-cosmonic): Mac builds can OOM in GitHub CI
+    continue-on-error: true
```
The only place where I have seen this fail was the install step, exactly once. Rather than silently ignoring errors here and completely breaking the assumptions of this CI check, the issue should be fixed in `build-nix` as was discussed offline - use plain bundled `install` binary rather than using GNU coreutils one.
If you don't want to do this in this PR, then please file an issue, but IMO, the proposed change is not a solution for the problem

vados-cosmonic replied:
And like we discussed offline, this is actually more complicated because `nix` did provide us a way to be cross-platform (in particular, we could ignore windows' lack of `install`) -- if you have some confidence that the error was only temporary/unlikely to reoccur, then I'll re-enable this step!

if not, I'd prefer to file an issue and have someone who is more experienced with `nix` take a look, and we can debug that way.

That said, there's another option here -- I'm going to mark `continue-on-error` only for *non* release branches. This way regular work can continue, but during a release, it has to pass.

vados-cosmonic commented:
rebased once more, fixed the release check ( to make it clear here, I'm expecting release branches [like this one](https://github.com/wasmCloud/wash/pull/614) to contain `release` somewhere (ex. `release/v0.18.0`) -- if this isn't standardized/process it won't work
